scipy/stats/stats.py

```diff
@@ -2465,15 +2465,20 @@ def gstd(a, axis=0, ddof=1):
             raise ValueError(
                 'Infinite value encountered. The geometric standard deviation '
                 'is defined for strictly positive values only.')
-        elif np.less_equal(a, 0).any():
+        a_nan = np.isnan(a)
+        a_nan_any = a_nan.any()
+        if a_nan.any():
+            return np.exp(np.std(log(a, where=~a_nan), axis=axis, ddof=ddof))
+        elif ((a_nan_any and np.less_equal(a[~a_nan], 0).any()) or
+              (not a_nan_any and np.less_equal(a, 0).any())):
             raise ValueError(
                 'Non positive value encountered. The geometric standard '
                 'deviation is defined for strictly positive values only.')
         elif 'Degrees of freedom <= 0 for slice' == str(w):
             raise ValueError(w)
         else:
             #  Remaining warnings don't need to be exceptions.
-            warnings.warn(w)
+            return np.exp(np.std(log(a, where=~a_nan), axis=axis, ddof=ddof))
```

Comment:
Just a quick confirmation; this avoids log dealing with the nan value that caused the spurious warning we're trying to avoid.

Reply:
Yes.

Note that in current master have reached this branch we reissue the warning and return None, which is unexpected. I was expecting the result of gstd instead. 

With this change, with or without use of `where`, an actual result would be returned.